prov/shm/src/smr_av.c

```diff
@@ -220,7 +220,8 @@ int smr_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
 	if (!smr_av)
 		return -FI_ENOMEM;
 
-	util_attr.addrlen = NAME_MAX;
+	util_attr.addrlen = NAME_MAX + 1;
```

Comment:
Sorry, did not realize that it's merged already.
This change looks like a hack, just to make the addrlen is multiple of 8 bytes. Do we really want to do this or is there something I'm missing?

Reply:
An assert was added to the util av code to check that the addrlen is a multiple of 8-bytes.  Though it likely only needs to be a multiple of sizeof(void *).  This was the easy change.

A more complex change would be to remove the addrlen size restriction, calculate an alignment offset for context_len, and align context that way.

Note that NAME_MAX excludes the null terminating character.  So adding 1 here actually makes some sense if the address includes it.  I can't easily figure out the intent of the shm code here.  In some cases, it uses NAME_MAX directly, in other cases it does NAME_MAX - 1.

Reply:
Yeah, I can instead add an additional field for 'addr_offset', that seems like a better solution

```
diff --git a/include/ofi_util.h b/include/ofi_util.h
index 0f57a851d..2870336e2 100644
--- a/include/ofi_util.h
+++ b/include/ofi_util.h
@@ -707,6 +708,7 @@ struct util_av {
        uint64_t                flags;
        size_t                  count;
        size_t                  addrlen;
+        size_t                  addr_offset;
        struct dlist_entry      ep_list;
        fastlock_t              ep_list_lock;
 };

diff --git a/prov/util/src/util_av.c b/prov/util/src/util_av.c
index eefd5cf0e..cd29fe9cf 100644
--- a/prov/util/src/util_av.c
+++ b/prov/util/src/util_av.c
@@ -269,7 +269,7 @@ int ofi_av_insert_addr(struct util_av *av, const void *addr, fi_addr_t *fi_addr)
 {
        struct util_av_entry *entry = NULL;

-       HASH_FIND(hh, av->hash, addr, av->addrlen, entry);
+       HASH_FIND(hh, av->hash, addr, av->addrlen + av->addr_offset, entry);
        if (entry) {
                if (fi_addr)
                        *fi_addr = ofi_buf_index(entry);
@@ -281,9 +281,9 @@ int ofi_av_insert_addr(struct util_av *av, const void *addr, fi_addr_t *fi_addr)
                        return -FI_ENOMEM;
                if (fi_addr)
                        *fi_addr = ofi_buf_index(entry);
-               memcpy(entry->data, addr, av->addrlen);
+               memcpy(entry->data, addr, av->addrlen + av->addr_offset);
                ofi_atomic_initialize32(&entry->use_cnt, 1);
-               HASH_ADD(hh, av->hash, data, av->addrlen, entry);
+               HASH_ADD(hh, av->hash, data, av->addrlen + av->addr_offset, entry);
        }
        return 0;
 }

@@ -423,10 +423,14 @@ static int util_av_init(struct util_av *av, const struct fi_av_attr *attr,
 {
        int ret = 0;
        size_t max_count;
+       size_t offset;

-       assert(util_attr->addrlen % 8 == 0);
+       /* offset calculated on a 8-byte boundary */
+       offset = util_attr->addrlen % 8;
+       if (offset != 0)`
+               offset = 8 - offset;
        struct ofi_bufpool_attr pool_attr = {
-               .size           = util_attr->addrlen +
+              .size           = util_attr->addrlen + offset +
                                  util_attr->context_len +
                                  sizeof(struct util_av_entry),
                .alignment      = 16,
@@ -449,6 +453,7 @@ static int util_av_init(struct util_av *av, const struct fi_av_attr *attr,
        FI_INFO(av->prov, FI_LOG_AV, "AV size %zu\n", av->count);
        av->addrlen = util_attr->addrlen;
+       av->addr_offset = offset;
```
Please let me know what you think about it

Reply:
The address should be aligned okay, since it's at the top of the data field.  The problem is that the context may be misaligned.  If the provider is expecting to cast &data[addrlen] to a structure, we need that aligned.  Switching to &data[context_offset] works.  We need to adjust the allocated data size based on extra alignment space.